stbt/android.py

```diff
@@ -0,0 +1,406 @@
+        except subprocess.CalledProcessError as e:
+            return e.output
+
+    def screenshot(self):
+        """Take a screenshot over adb, in the same orientation that the device
+        is in."""
+
+        for attempt in range(1, 11):
+            data = (self.adb(["shell", "screencap", "-p"],
+                             timeout_secs=60, capture_output=True)
+                    .replace("\r\n", "\n"))
+            img = cv2.imdecode(
+                numpy.asarray(bytearray(data), dtype=numpy.uint8),
+                cv2.CV_LOAD_IMAGE_COLOR)
+            if img is None:
+                sys.stderr.write(
+                    "AdbDevice.screenshot: error: Failed to get screenshot "
+                    "via adb (attempt %d/10)\n" % attempt)
+                sys.stderr.write("Length of data: %d\n" % len(data))
+                sys.stderr.write("Data: %r\n" % data)
+            else:
+                break
+        else:
+            raise RuntimeError(
+                "Failed to capture screenshot from android device")
+
+        w, h = img.shape[1], img.shape[0]
+        stbt.debug("Screenshot is %d px wide x %d px high" % (w, h))
+
+        if self.coordinate_system == CoordinateSystem.ADB_720P:
+            # Resize to 720p preserving orientation
+            if w > h:
+                img = cv2.resize(img, (1280, 720))
+            else:
+                img = cv2.resize(img, (720, 1280))
+
+        return img
+
+    def swipe(self, start_position, end_position):
+        """Swipe from one point to another point.
+
+        :param start_position:
+            A `stbt.Region` or (x, y) tuple of coordinates at which to start.
+        :param end_position:
+            A `stbt.Region` or (x, y) tuple of coordinates at which to stop.
+
+        Example::
+
+          swipe((100, 100), (100, 400))
+
+        """
+        x1, y1 = _region_to_tuple(start_position)
+        x2, y2 = _region_to_tuple(end_position)
+        stbt.debug("AdbDevice.swipe((%d,%d), (%d,%d))" % (x1, y1, x2, y2))
+
+        x1, y1 = self._to_native_coordinates(x1, y1)
+        x2, y2 = self._to_native_coordinates(x2, y2)
+        command = ["shell", "input", "swipe",
+                   str(x1), str(y1), str(x2), str(y2)]
+        self.adb(command, timeout_secs=10)
+
+    def tap(self, position):
+        """Tap on a particular location.
+
+        :param position: A `stbt.Region`, or an (x,y) tuple.
+
+        Example::
+
+            tap((100, 20))
+            tap(stbt.match(...).region)
+
+        """
+        x, y = _region_to_tuple(position)
+        stbt.debug("AdbDevice.tap((%d,%d))" % (x, y))
+
+        x, y = self._to_native_coordinates(x, y)
+        self.adb(["shell", "input", "tap", str(x), str(y)], timeout_secs=10)
+
+    def _adb(self, command, timeout_secs=None, **kwargs):
+        _command = []
+        if timeout_secs is not None:
+            _command += ["timeout", "%fs" % timeout_secs]
```

Comment:
Hmmm *all* of the subprocess functions have `timeout` since Python 3.3, and the `subprocess32` backport includes this. Unfortunately `subprocess32` isn't available in the Ubuntu 12.04 or 14.04 repos.